Tutorial: Create an Automated Market Maker (#2113)
XRPLF/xrpl-dev-portal: content/tutorials/use-tokens/create-an-automated-market-maker.md

---
parent: use-tokens.html
blurb: Set up an Automated Market Maker (AMM)
embed_xrpl_js: true
status: not_enabled
filters:
  - interactive_steps
labels:
  - Decentralized Exchange
  - Tokens
  - AMM
---
# Create an Automated Market Maker
_(Requires the [AMM amendment][] :not_enabled:)_

> An [Automated Market Maker (AMM)](automated-market-makers.html) can be an efficient way to facilitate exchanges between two assets, while earning its liquidity providers passive income. This tutorial shows how to create the AMM for a given asset pair.

Reviewer: ```suggestion
An [Automated Market Maker (AMM)](automated-market-makers.html) can be an efficient way to facilitate exchanges between two assets while earning its liquidity providers passive income. This tutorial shows how to create an AMM for a given asset pair.
```

<!-- Source for this specific tutorial's interactive bits: -->
<script type="application/javascript" src="assets/js/tutorials/create-amm.js"></script>

## Prerequisites

> - You must have an XRP Ledger address and some XRP. For testing, you can get these from a [Faucet](xrp-testnet-faucet.html).

Reviewer: ```suggestion
- You must have an XRP Ledger address and some XRP. For development and testing purposes, you can get these from a [Faucet](xrp-testnet-faucet.html).
```

- You should have a basic understanding of how [tokens](tokens.html) work in the XRP Ledger.
- You may want to read about [Automated Market Makers in the XRP Ledger](automated-market-makers.html) first.

## Example Code

Complete sample code for all of the steps of these tutorials is available under the [MIT license](https://github.com/XRPLF/xrpl-dev-portal/blob/master/LICENSE).

- See [Code Samples: Create an AMM](https://github.com/XRPLF/xrpl-dev-portal/tree/master/content/_code-samples/create-amm/) in the source repository for this website.

## Steps
{% set n = cycler(* range(1,99)) %}
{% set use_network = "Devnet" %}

### {{n.next()}}. Connect to the network

You must be connected to the network to query it and submit transactions. The following code shows how to connect to a public {{use_network}} server using a supported [client library](client-libraries.html):

<!-- MULTICODE_BLOCK_START -->

_JavaScript_

{{ include_code("_code-samples/create-amm/js/connect.js", language="js") }}

<!-- MULTICODE_BLOCK_END -->

For this tutorial, click the following button to connect:

{% include '_snippets/interactive-tutorials/connect-step.md' %}

### {{n.next()}}. Get credentials

> To transact on the XRP Ledger, you need an address and secret key, and some XRP. For development purposes, you can get these on the [{{use_network}}](parallel-networks.html) using the following interface:

Reviewer: ```suggestion
To transact on the XRP Ledger, you need an address, a secret key, and some XRP. For development and testing purposes, you can get these on the [{{use_network}}](parallel-networks.html) using the following interface:
```

{% include '_snippets/interactive-tutorials/generate-step.md' %}

When you're [building production-ready software](production-readiness.html), you should use an existing account, and manage your keys using a [secure signing configuration](set-up-secure-signing.html). The following code shows how to get a `Wallet` instance using either the faucet or a seed provided by environment variable:

<!-- MULTICODE_BLOCK_START -->

_JavaScript_

{{ include_code("_code-samples/create-amm/js/create-amm.js", language="js", start_with="// Get credentials", end_before="// Acquire tokens") }}

<!-- MULTICODE_BLOCK_END -->

### {{n.next()}}. Select and acquire assets

> As the creator of an AMM, you are also the first liquidity provider and you have to supply it with a starting pool of assets. Other users of the XRP Ledger can also become liquidity providers by supplying assets after the AMM exists. It's important to choose assets especially carefully because, as a liquidity provider for an AMM, you are supplying some amounts of both for users to swap between. If one of the AMM's assets becomes worthless, other users can use the AMM to trade for the other asset, leaving the AMM (and thus, its liquidity providers including you) holding only the worthless one. Technically, the AMM always holds some positive amount of both assets, but the amounts can be very small.

Reviewer: ```suggestion
As the creator of an AMM, you are also the first liquidity provider and you have to supply it with a starting pool of assets. Other users of the XRP Ledger can also become liquidity providers by supplying assets after the AMM exists. It's crucial to choose assets carefully because, as a liquidity provider for an AMM, you are supplying some amounts of both for users to swap between. If one of the AMM's assets becomes worthless, other users can use the AMM to trade for the other asset, leaving the AMM (and thus, its liquidity providers including you) holding only the worthless one. Technically, the AMM always holds some positive amount of both assets, but the amounts can be very small.
```

<!-- MULTICODE_BLOCK_START -->

_JavaScript_

{{ include_code("_code-samples/create-amm/js/create-amm.js", language="js", start_with="/* Issue tokens") }}

<!-- MULTICODE_BLOCK_END -->

{{ start_step("Acquire tokens") }}
<button id="buy-tst" class="btn btn-primary previous-steps-required">Buy TST</button>
<button id="get-foo" class="btn btn-primary previous-steps-required">Get FOO</button>
<div class="loader collapse"><img class="throbber" src="assets/img/xrp-loader-96.png">Working...</div>
<div class="output-area"></div>
{{ end_step() }}

### {{n.next()}}. Check if the AMM exists

Since there can only be one AMM for a specific pair of assets, it's best to check first before trying to create one. Use the [amm_info method][] to check whether the AMM already exists. For the request, you specify the two assets. The response should be an `actNotFound` error if the AMM does not exist.

<!-- MULTICODE_BLOCK_START -->

_JavaScript_

{{ include_code("_code-samples/create-amm/js/create-amm.js", language="js", start_with="// Check if AMM already exists", end_before="// Look up AMM transaction cost") }}

<!-- MULTICODE_BLOCK_END -->

> If the AMM does already exist, you should double-check that you specified the right pair of assets; if so, someone else has already created this AMM, but you can still deposit to it instead. <!-- TODO: link to a tutorial about depositing to and withdrawing from an AMM when one exists -->

Reviewer: ```suggestion
If the AMM does already exist, you should double-check that you specified the right pair of assets. If someone else has already created this AMM, you can deposit to it instead. <!-- TODO: link to a tutorial about depositing to and withdrawing from an AMM when one exists -->
```

{{ start_step("Check for AMM") }}
<button id="check-for-amm" class="btn btn-primary previous-steps-required">Check AMM</button>
<div class="loader collapse"><img class="throbber" src="assets/img/xrp-loader-96.png">Sending...</div>
<div class="output-area"></div>
{{ end_step() }}

### {{n.next()}}. Look up the AMMCreate transaction cost

Creating an AMM has a special [transaction cost][] to prevent spam: since it creates objects in the ledger that no one owns, you must burn at least one [owner reserve increment](reserves.html) of XRP to send the AMMCreate transaction. The exact value can change due to [fee voting](https://xrpl.org/fee-voting.html), so you should look up the current incremental reserve value using the [server_state method][].

> It is also a good practice to display this value and give a human operator a chance to stop before you send the transaction. Burning an owner reserve is typically a much higher cost than sending a normal transaction, so you don't want it to be a surprise. (Currently, on both Mainnet and Devnet, the cost of sending a typical transaction is 0.000010 XRP but the cost of AMMCreate is 2 XRP.)

Reviewer: I'd avoid using "Currently" to note the cost of AMMCreate as it could become outdated. Maybe "In September 2023"?

Author: We have several places in the docs that say something like this. I'd rather have it say, "Currently," and have to go update those when the numbers change, than to say, "As of September 2023," and have to constantly update it—if the date given is a year old, it may leave uncertainty about whether it's still accurate.

Transaction costs and reserves change rarely enough that I don't think it'll be a problem. The last change was in 2021 and before that they hadn't changed since before 2014. (They changed a couple times, I think, in the very early days, but I wasn't around for that.)
